File: apps/frontend/app/src/main/java/com/nomadclub/cashchat/config/AppConfig.kt
package com.nomadclub.cashchat.config

import com.nomadclub.cashchat.BuildConfig

/**
 * 앱 전역 설정 진입점(세션 스냅샷).
 *
 * 계층형 폴백으로 값을 해석한다:
 * `Remote Config(활성값) → BuildConfig(빌드타임) → Google 테스트 ID`
 *
 * 광고/정책 값은 세션 시작 시 1회만 읽어 세션 도중 변하지 않는다(다음 실행에 반영).
 * 긴급 키(점검/강제업데이트)만 [RemoteConfigManager.gateState]로 reactive 하게 처리.
 *
 * > AdMob **앱 ID**는 Manifest/SDK 초기화에 필요해 빌드타임 전용 — RC로 교체 불가.
 */
data class AppConfig(
    // API
    val baseUrl: String,
    // AdMob
    val adsEnabled: Boolean,
    val admobAppId: String,
    val admobBannerAdUnitId: String,
    val admobInterstitialAdUnitId: String,
    val admobNativeAdUnitId: String,
    val admobRewardedAdUnitId: String,
    // Sentry
    val sentryDsn: String,
    // TNK Offerwall
    val tnkAppId: String,
    // Feature flags / 정책 (Remote Config)
    val offerwallEnabled: Boolean,
    val adChatInterval: Long,
    val rewardChatInterval: Long,
    val rewardRequired: Boolean,
    val interstitialTriggerAction: String,
) {
    companion object {
        /** Remote Config 미사용 환경(테스트 등)의 폴백 생성자: BuildConfig→테스트ID. */
        fun fromBuildConfig(): AppConfig = AppConfig(
            baseUrl = BuildConfig.BASE_URL,
            adsEnabled = true,
            admobAppId = BuildConfig.ADMOB_APP_ID.ifBlank { TestAdUnitIds.APP_ID },
            admobBannerAdUnitId = BuildConfig.ADMOB_BANNER_AD_UNIT_ID.ifBlank { TestAdUnitIds.BANNER },
            admobInterstitialAdUnitId = BuildConfig.ADMOB_INTERSTITIAL_AD_UNIT_ID.ifBlank { TestAdUnitIds.INTERSTITIAL },
            admobNativeAdUnitId = BuildConfig.ADMOB_NATIVE_AD_UNIT_ID.ifBlank { TestAdUnitIds.NATIVE },
            admobRewardedAdUnitId = BuildConfig.ADMOB_REWARDED_AD_UNIT_ID.ifBlank { TestAdUnitIds.REWARDED },
            sentryDsn = BuildConfig.SENTRY_DSN,
            tnkAppId = BuildConfig.TNK_APP_ID,
            offerwallEnabled = true,
            adChatInterval = 1L,
            rewardChatInterval = 3L,
            rewardRequired = true,
            interstitialTriggerAction = "new_chat",
        )

        /** RC(활성값)→BuildConfig→테스트ID 계층 폴백으로 해석. */
        fun resolve(rc: RemoteConfigManager): AppConfig = AppConfig(
            baseUrl = BuildConfig.BASE_URL,
            adsEnabled = rc.getBoolean(RemoteConfigKeys.ADS_ENABLED),
            // 앱 ID는 빌드타임 전용
            admobAppId = BuildConfig.ADMOB_APP_ID.ifBlank { TestAdUnitIds.APP_ID },
            admobBannerAdUnitId = adUnit(rc, RemoteConfigKeys.ADMOB_BANNER_AD_UNIT_ID, BuildConfig.ADMOB_BANNER_AD_UNIT_ID, TestAdUnitIds.BANNER),
            admobInterstitialAdUnitId = adUnit(rc, RemoteConfigKeys.ADMOB_INTERSTITIAL_AD_UNIT_ID, BuildConfig.ADMOB_INTERSTITIAL_AD_UNIT_ID, TestAdUnitIds.INTERSTITIAL),
            admobNativeAdUnitId = adUnit(rc, RemoteConfigKeys.ADMOB_NATIVE_AD_UNIT_ID, BuildConfig.ADMOB_NATIVE_AD_UNIT_ID, TestAdUnitIds.NATIVE),
            admobRewardedAdUnitId = adUnit(rc, RemoteConfigKeys.ADMOB_REWARDED_AD_UNIT_ID, BuildConfig.ADMOB_REWARDED_AD_UNIT_ID, TestAdUnitIds.REWARDED),
            sentryDsn = BuildConfig.SENTRY_DSN,
            tnkAppId = BuildConfig.TNK_APP_ID,
            offerwallEnabled = rc.getBoolean(RemoteConfigKeys.OFFERWALL_ENABLED),
            adChatInterval = rc.getLong(RemoteConfigKeys.AD_CHAT_INTERVAL),
            rewardChatInterval = rc.getLong(RemoteConfigKeys.REWARD_CHAT_INTERVAL),
            rewardRequired = rc.getBoolean(RemoteConfigKeys.REWARD_REQUIRED),
            interstitialTriggerAction = rc.getString(RemoteConfigKeys.INTERSTITIAL_TRIGGER_ACTION),
        )

        private fun adUnit(rc: RemoteConfigManager, key: String, buildTime: String, test: String): String =
            rc.getString(key).ifBlank { buildTime.ifBlank { test } }
    }
}

File: apps/frontend/CashChatIOS/CashChatIOS/AppConfig.swift
import Foundation

/// 앱 전역 설정 진입점(세션 스냅샷).
///
/// 광고/정책 값은 계층형 폴백으로 해석한다:
/// `Remote Config(활성값) → Secrets(빌드타임) → Google 테스트 ID`
/// 정적 프로퍼티라 최초 접근 시 1회만 계산되어 세션 도중 변하지 않는다(다음 실행 반영).
///
/// > AdMob **앱 ID**는 Info.plist/SDK 초기화에 필요해 빌드타임 전용 — RC로 교체 불가.
enum AppConfig {
    static let googleIOSClientId: String = required(Secrets.googleIOSClientId, key: "googleIOSClientId")
    static let googleWebClientId: String = required(Secrets.googleWebClientId, key: "googleWebClientId")
    static let apiBaseUrl: String = required(Secrets.apiBaseUrl, key: "apiBaseUrl")
    static let tnkAppId: String = required(Secrets.tnkAppId, key: "tnkAppId")

    // AdMob — RC→Secrets→테스트ID 폴백
    static let adsEnabled: Bool = RemoteConfigManager.shared.bool(RemoteConfigKeys.adsEnabled)
    static let admobAppId: String = fallback(Secrets.admobAppId, test: TestAdUnitIds.appId)
    static let admobRewardedAdUnitId: String = adUnit(RemoteConfigKeys.admobRewardedAdUnitId, Secrets.admobRewardedAdUnitId, TestAdUnitIds.rewarded)
    static let admobBannerAdUnitId: String = adUnit(RemoteConfigKeys.admobBannerAdUnitId, Secrets.admobBannerAdUnitId, TestAdUnitIds.banner)
    static let admobInterstitialAdUnitId: String = adUnit(RemoteConfigKeys.admobInterstitialAdUnitId, Secrets.admobInterstitialAdUnitId, TestAdUnitIds.interstitial)
    static let admobNativeAdUnitId: String = adUnit(RemoteConfigKeys.admobNativeAdUnitId, Secrets.admobNativeAdUnitId, TestAdUnitIds.native)

    // Feature flags / 정책 (Remote Config)
    static let offerwallEnabled: Bool = RemoteConfigManager.shared.bool(RemoteConfigKeys.offerwallEnabled)
    // RC 오입력(0/음수) 방어 — 간격은 최소 1 이상이어야 한다(modulo/카운트 계산 안전).
    static let adChatInterval: Int = max(1, RemoteConfigManager.shared.number(RemoteConfigKeys.adChatInterval).intValue)
    static let rewardChatInterval: Int = max(1, RemoteConfigManager.shared.number(RemoteConfigKeys.rewardChatInterval).intValue)
    static let rewardRequired: Bool = RemoteConfigManager.shared.bool(RemoteConfigKeys.rewardRequired)
    static let interstitialTriggerAction: String = RemoteConfigManager.shared.string(RemoteConfigKeys.interstitialTriggerAction)

    /// RC→Secrets→테스트ID 폴백.
    private static func adUnit(_ rcKey: String, _ secret: String, _ test: String) -> String {
        let rc = RemoteConfigManager.shared.string(rcKey).trimmingCharacters(in: .whitespacesAndNewlines)
        if !rc.isEmpty { return rc }
        return fallback(secret, test: test)
    }

    /// Secrets 값이 비었거나 placeholder면 테스트 ID로 폴백(앱 크래시 방지).
    private static func fallback(_ secret: String, test: String) -> String {
        let trimmed = secret.trimmingCharacters(in: .whitespacesAndNewlines)
        if !trimmed.isEmpty, !trimmed.hasPrefix("YOUR_") { return trimmed }
        return test
    }

    /// 필수 비밀값. 비었거나 placeholder면 즉시 실패(개발 중 설정 누락 조기 발견).
    private static func required(_ value: String, key: String) -> String {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, !trimmed.hasPrefix("YOUR_") else {
            preconditionFailure(
                "[\(key)] 값이 설정되지 않았습니다.\n" +
                "Secrets.swift를 생성하고 실제 값을 입력하세요."
            )
        }
        return trimmed
    }
}
